source/core/slang-shader-cache-index.cpp

```diff
@@ -0,0 +1,173 @@
+    {
+        List<UnownedStringSlice> digests;
+        StringUtil::split(line, ' ', digests); // This will return our two hashes as two elements in digests.
+        auto dependencyDigest = stringToHash(digests[0]);
+        auto astDigest = stringToHash(digests[1]);
+
+        ShaderCacheEntry entry = { dependencyDigest, astDigest };
+        auto entryNode = entries.AddFirst(entry);
+        keyToEntry.Add(dependencyDigest, entryNode);
+    }
+}
+
+LinkedNode<ShaderCacheEntry>* ShaderCacheIndex::findEntry(const slang::Digest& key, ISlangBlob** outCompiledCode)
+{
+    LinkedNode<ShaderCacheEntry>* entryNode;
+    if (!keyToEntry.TryGetValue(key, entryNode))
+    {
+        // The key was not found in the cache, so we return nullptr.
+        *outCompiledCode = nullptr;
+        return nullptr;
+    }
+
+    // If the key is found, we need to move its corresponding entry to the front of
+    // the list and load the stored contents from disk.
+    entries.RemoveFromList(entryNode);
+    entries.AddFirst(entryNode);
+    shaderCacheFileSystem->loadFile(hashToString(key).getBuffer(), outCompiledCode);
+    return entryNode;
+}
+
+void ShaderCacheIndex::addEntry(const slang::Digest& dependencyDigest, const slang::Digest& astDigest, ISlangBlob* compiledCode)
+{
+    if (!mutableShaderCacheFileSystem)
+    {
+        // Should not save new entries if the underlying file system isn't mutable.
+        return;
+    }
+    
+    // Check that we do not exceed the cache's size limit by adding another entry. If so,
+    // remove the least recently used entry first.
+    while (entryCountLimit > 0 && entries.Count() >= entryCountLimit)
+    {
+        deleteLRUEntry();
+    }
+
+    ShaderCacheEntry entry = { dependencyDigest, astDigest };
+    auto entryNode = entries.AddFirst(entry);
+    keyToEntry.Add(dependencyDigest, entryNode);
+
+    mutableShaderCacheFileSystem->saveFileBlob(hashToString(dependencyDigest).getBuffer(), compiledCode);
+
+    saveCacheIndexToFile();
+}
+
+void ShaderCacheIndex::updateEntry(
+    LinkedNode<ShaderCacheEntry>* entryNode,
+    const slang::Digest& dependencyDigest,
+    const slang::Digest& astDigest,
+    ISlangBlob* updatedCode)
+{
+    if (!mutableShaderCacheFileSystem)
+    {
+        // Updating entries requires saving to disk in order to overwrite the old shader file
+        // on disk, so we return if the underlying file system isn't mutable.
+        return;
+    }
+
+    entryNode->Value.astBasedDigest = astDigest;
+    mutableShaderCacheFileSystem->saveFileBlob(hashToString(dependencyDigest).getBuffer(), updatedCode);
+
+    saveCacheIndexToFile();
+}
+
+void ShaderCacheIndex::saveCacheIndexToFile()
+{
+    if (!mutableShaderCacheFileSystem)
+    {
+        // Cannot save the index to disk if the underlying file system isn't mutable.
+        return;
+    }
+
+    StringBuilder indexString;
+    for (auto& entry : entries)
+    {
+        indexString.append(hashToString(entry.dependencyBasedDigest));
```

Comment:
To speed things up we should probably have an operator override in `digest-util` for `StringBuilder`:
```
StringBuilder& operator <<(StringBuilder& sb, const Digest& d);
```
Then just call `indexSb << entry.dependencyBasedDigest;` here to avoid the allocation involved in creating the temporary `String`s.